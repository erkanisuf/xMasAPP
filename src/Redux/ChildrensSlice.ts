import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { IProduct } from "../components/ChildComp/ChildComp";

import type { RootState } from "./store";

interface IReduxChildrenSlice {
  ChildrenApprovedItems: OBJtoAPI[];
  ChildrenDiscardedItems: OBJtoAPI[];
  myCart: IProduct[];
}
export interface OBJtoAPI {
  userId: number;
  date: Date | string;
  products: IProduct[];
}
// This is the Initial State represents the Products from the Open APi
const initialState: IReduxChildrenSlice = {
  ChildrenApprovedItems: [],
  ChildrenDiscardedItems: [],
  myCart: [],
};

export const ChildrensSlice = createSlice({
  name: "childrenSlice",
  //Redux Info
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    //Redux Tips
    // Use the PayloadAction type to declare the contents of `action.payload`
    ChildProductToApproved: (state, action: PayloadAction<OBJtoAPI>) => {
      const copyApprovedState = [...state.ChildrenApprovedItems];
      const findIndex = copyApprovedState.findIndex(
        (el) => el.userId === action.payload.userId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyApprovedState[findIndex].products.push(action.payload.products[0]);
      } else {
        copyApprovedState.push(action.payload);
      }

      state.ChildrenApprovedItems = copyApprovedState;
    },
    ChildProductToDiscarded: (state, action: PayloadAction<OBJtoAPI>) => {
      const copyDiscardedState = [...state.ChildrenDiscardedItems];
      const findIndex = copyDiscardedState.findIndex(
        (el) => el.userId === action.payload.userId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyDiscardedState[findIndex].products.push(action.payload.products[0]);
      } else {
        copyDiscardedState.push(action.payload);
      }

      state.ChildrenDiscardedItems = copyDiscardedState;
    },
    AddItemToMyCart: (state, action: PayloadAction<IProduct>) => {
      const copyMyCart = [...state.myCart];
      const product = { productId: action.payload.productId, quantity: 1 }; //New Object because the APi has random values of quantity
      const findIndex = copyMyCart.findIndex(
        (el) => el.productId === action.payload.productId
      );
      //If the user is already there , it finds its index and just updates the products array.
      if (findIndex !== -1) {
        copyMyCart[findIndex].quantity++;
      } else {
        copyMyCart.push(product);
      }

      state.myCart = copyMyCart;
    },
  },
});

export const {
  ChildProductToApproved,
  ChildProductToDiscarded,
  AddItemToMyCart,
} = ChildrensSlice.actions;

//Redux Setup from their Docs.
// Other code such as selectors can use the imported `RootState` type
export const selectCount = (state: RootState) => state;

export default ChildrensSlice.reducer;
