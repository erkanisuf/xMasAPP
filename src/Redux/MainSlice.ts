import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { IProductDesc } from "../components/Product/Product";
import type { RootState } from "./store";

interface IReduxMainSlice {
  products: IProductDesc[];
  myChildren: { id: number; name: string; image: string }[];
}
interface IChangePrice {
  id: number;
  discount: number;
}

// This is the Initial State represents the Products from the Open APi
const initialState: IReduxMainSlice = {
  products: [],
  myChildren: [
    {
      id: 1,
      name: "Pekka",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSVY3efFyktSBShjhBcNUt2JuH3F1r0qnlt1g&usqp=CAU",
    },
    {
      id: 2,
      name: "Mikko",
      image:
        "https://www.pngitem.com/pimgs/m/424-4240458_one-ok-rock-stand-out-fit-in-from.png",
    },
    {
      id: 3,
      name: "Juha",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRdBkq9ItXGCTqTIW4BdBlaEpZQVllKv8gkQGpdhVwfbAIuj9-4yOzapTjdNE0r14wSJoE&usqp=CAU",
    },
    {
      id: 4,
      name: "Jonna",
      image:
        "https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQlCT3cJxdz3-ntu8ixFRexArdaknweGrBQrw&usqp=CAU",
    },
    {
      id: 5,
      name: "Erkku",
      image:
        "https://st4.depositphotos.com/18690434/20926/v/450/depositphotos_209266078-stock-illustration-boy-kid-avatar-icon-vector.jpg",
    },
  ],
};

export const MainSlice = createSlice({
  name: "main",
  //Redux Info
  // `createSlice` will infer the state type from the `initialState` argument
  initialState,
  reducers: {
    //Redux Tips
    // Use the PayloadAction type to declare the contents of `action.payload`
    addProductsFromApi: (state, action: PayloadAction<IProductDesc[]>) => {
      state.products = action.payload;
    },
    changeProductPrice: (state, action: PayloadAction<IChangePrice>) => {
      const copyProducts = [...state.products];
      const findIndex = copyProducts.findIndex(
        (el) => el.id === action.payload.id
      );
      //Discount if more than 1 quantity
      if (findIndex !== -1) {
        copyProducts[findIndex] = {
          ...copyProducts[findIndex],
          discount: action.payload.discount,
        };
      }
      state.products = copyProducts;
    },
  },
});

export const { addProductsFromApi, changeProductPrice } = MainSlice.actions;

//Redux Setup from their Docs.
// Other code such as selectors can use the imported `RootState` type
export const selectCount = (state: RootState) => state;

export default MainSlice.reducer;
