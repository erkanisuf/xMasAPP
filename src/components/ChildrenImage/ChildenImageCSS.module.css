.containerChild {
  display: flex;
  justify-content: center;
  align-items: center;
  border-bottom: 1px solid #ccc;
  padding: 10px;
  margin: 25px auto;
  width: 80%;
}
.containerChild img {
  width: 5%;
  margin: 0 15px;
  border-radius: 100%;
  border: 5px solid #ccc;
  background-color: white;
}
