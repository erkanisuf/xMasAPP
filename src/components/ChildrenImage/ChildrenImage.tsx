import React, { useMemo, useState } from "react";
import { useAppSelector } from "../../Redux/hooks";
import ChildenImageCSS from "./ChildenImageCSS.module.css";
interface IChildrenImage {
  name: number; // This is The ID
}

interface IChild {
  id: number;
  name: string;
  image: string;
}
const ChildrenImage: React.FC<IChildrenImage> = ({ name }) => {
  //Component that shows image and name of children
  const myChildren = useAppSelector((state) => state.main.myChildren); // Redux Main State
  const [child, setChild] = useState<IChild>();

  //finds children information form the Redux
  useMemo(() => {
    const findChild = myChildren.find((el) => el.id === name); // this finds child from Redux by ID
    setChild(findChild);
    return () => {};
  }, [name, myChildren]);

  if (!child) {
    return <p>loading</p>;
  }
  return (
    <div className={ChildenImageCSS.containerChild}>
      <img src={child.image} alt={""} />
      <h1>{child.name}`s wishlist </h1>
    </div>
  );
};

export default ChildrenImage;
