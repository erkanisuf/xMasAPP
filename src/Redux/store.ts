import { configureStore } from "@reduxjs/toolkit";
import ChildrensSlice from "./ChildrensSlice";
import MainSlice from "./MainSlice";
// ...

//MAIN: holds the products and the fake array of childrens object
// CHILDREN : holds the approved items and discarded items of the child , laso myCart property that holds the item that customer will  "buy"
export const store = configureStore({
  reducer: { main: MainSlice, childrens: ChildrensSlice },
});

// Redux required setup from Their Docs
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
