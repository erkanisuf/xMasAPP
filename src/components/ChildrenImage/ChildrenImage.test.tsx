import React from "react";
import { Provider } from "react-redux";

import renderer from "react-test-renderer";
import { store } from "../../Redux/store";
import ChildrenImage from "./ChildrenImage";

it("snapshot <ChildrenImage/> Component", () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <ChildrenImage name={1} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
