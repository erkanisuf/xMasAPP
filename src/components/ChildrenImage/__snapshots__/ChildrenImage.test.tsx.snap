// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot <ChildrenImage/> Component 1`] = `
<div
  className="containerChild"
>
  <img
    alt=""
    src="https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSVY3efFyktSBShjhBcNUt2JuH3F1r0qnlt1g&usqp=CAU"
  />
  <h1>
    Pekka
    \`s wishlist 
  </h1>
</div>
`;
